File: ediparser/parser/schema.py
"""
Pydantic schemas for parsed EDI data
Maps X12 segments to a canonical JSON structure

Field names here are the contract the api-gateway reads in
`_clean_claim` / `_clean_denial` (api-gateway/routes/ingestion.py).
Renaming anything below means changing that file too.
"""

from typing import Optional

from pydantic import BaseModel, Field


# -- Payment Summary --
class ParsedPaymentInfo(BaseModel):
    """BPR segment - Beginning Segment for Payment Order/Remittance Advice"""
    transaction_handling_code: Optional[str] = None   # BPR01: I, H, C, D, ...
    payment_amount: float = 0.0                       # BPR02
    credit_debit_flag: Optional[str] = None           # BPR03: C or D
    payment_method: Optional[str] = None              # BPR04: CHK, ACH, FWT, NON, BOP
    payment_format: Optional[str] = None              # BPR05: CCP, CTX
    payment_date: Optional[str] = None                # BPR16 check/EFT effective date
    trace_number: Optional[str] = None                # TRN02
    payer_identifier: Optional[str] = None            # TRN03
    production_date: Optional[str] = None             # DTM*405
    payer_name: Optional[str] = None                  # N1*PR
    payer_id_number: Optional[str] = None
    payee_name: Optional[str] = None                  # N1*PE
    payee_npi: Optional[str] = None                   # N1*PE with XX qualifier


# -- Claim Adjustment --
class ParsedClaimAdjustment(BaseModel):
    """One adjustment triplet from a CAS segment"""
    adjustment_group_code: str                # CAS01: PR, CO, OA, PI (AB/AS legacy)
    reason_code: Optional[str] = None         # CARC
    amount: float = 0.0
    quantity: Optional[float] = None
    remark_code: Optional[str] = None         # RARC, sourced from LQ / MOA / MIA
    description: Optional[str] = None
    level: str = "claim"                      # "claim" or "line"


# -- Service Line --
class ParsedServiceLine(BaseModel):
    """SVC segment - Service Payment Information"""
    line_number: int
    procedure_qualifier: Optional[str] = None   # SVC01-1: HC, HP, IV, ER, ...
    cpt_code: Optional[str] = None
    hcpcs_code: Optional[str] = None
    modifier_1: Optional[str] = None
    modifier_2: Optional[str] = None
    modifier_3: Optional[str] = None
    modifier_4: Optional[str] = None
    charge_amount: float = 0.0                  # SVC02
    paid_amount: float = 0.0                    # SVC03
    revenue_code: Optional[str] = None          # SVC04
    units: float = 1.0                          # SVC05
    allowed_amount: Optional[float] = None      # AMT*B6
    service_date: Optional[str] = None          # DTM*472
    control_number: Optional[str] = None        # REF*6R
    remark_codes: list[str] = Field(default_factory=list)


# -- Service Line with Adjustments --
class ParsedServiceLineDetail(BaseModel):
    """Combined service line with its adjustments"""
    service_line: ParsedServiceLine
    adjustments: list[ParsedClaimAdjustment] = Field(default_factory=list)


# -- Claim --
class ParsedClaim(BaseModel):
    """CLP segment - Claim Payment Information, plus its loop"""
    claim_id: str                                     # CLP01 patient control number
    claim_status_code: str                            # CLP02: 1 paid, 4 denied, 22 reversal
    total_charged: float = 0.0                        # CLP03
    total_paid: float = 0.0                           # CLP04
    patient_responsibility: float = 0.0               # CLP05
    total_adjustment: float = 0.0                     # summed from CAS, not read from CLP
    claim_filing_indicator: Optional[str] = None      # CLP06
    payer_claim_control_number: Optional[str] = None  # CLP07
    facility_type_code: Optional[str] = None          # CLP08
    frequency_code: Optional[str] = None              # CLP09
    claim_type: str = "professional"                  # DB CHECK: professional/institutional/pharmacy

    patient_id: Optional[str] = None                  # NM1*QC identifier
    patient_name: Optional[str] = None                # NM1*QC name
    date_of_birth: Optional[str] = None               # DMG02 when present
    provider_npi: Optional[str] = None                # NM1*82 rendering provider
    provider_name: Optional[str] = None
    payer_name: Optional[str] = None                  # carried down from the N1*PR loop
    payer_id_number: Optional[str] = None

    service_from: Optional[str] = None                # DTM*232 or earliest line date
    service_to: Optional[str] = None                  # DTM*233 or latest line date
    diagnosis_codes: list[str] = Field(default_factory=list)

    service_lines: list[ParsedServiceLineDetail] = Field(default_factory=list)
    claim_level_adjustments: list[ParsedClaimAdjustment] = Field(default_factory=list)
    remark_codes: list[str] = Field(default_factory=list)


# -- Denial (derived from claim data) --
class ParsedDenial(BaseModel):
    """Denial record derived from a CAS adjustment"""
    claim_id: str
    service_line_number: Optional[int] = None
    cpt_code: Optional[str] = None
    hcpcs_code: Optional[str] = None
    modifier_1: Optional[str] = None
    modifier_2: Optional[str] = None
    charge_amount: float = 0.0
    payment_amount: float = 0.0
    adjustment_amount: float = 0.0
    cagc: str                                   # CO, PR, OA, PI - DB CHECK constrains this
    carc_code: Optional[str] = None
    rarc_code: Optional[str] = None
    denial_reason: Optional[str] = None
    denial_date: Optional[str] = None


# -- Transaction Metadata --
class ParsedTransactionMetadata(BaseModel):
    """ISA/GS/ST envelope info"""
    transaction_set_identifier: Optional[str] = None            # ST01: 835 or 837
    transaction_set_control_number: Optional[str] = None        # ST02
    implementation_convention_reference: Optional[str] = None   # ST03
    interchange_sender_id: Optional[str] = None                 # ISA06
    interchange_receiver_id: Optional[str] = None               # ISA08
    functional_group_code: Optional[str] = None                 # GS01
    version_identifier: Optional[str] = None                    # GS08
    element_separator: Optional[str] = None
    component_separator: Optional[str] = None
    segment_terminator: Optional[str] = None


# -- Full Parsed Response --
class Parsed835Response(BaseModel):
    """Complete parsed EDI response (835 remittance or 837 claim submission)"""
    metadata: ParsedTransactionMetadata
    payment_info: ParsedPaymentInfo
    claims: list[ParsedClaim]
    denials: list[ParsedDenial]
    total_claims: int
    total_denials: int
    total_charges: float
    total_payments: float
    total_adjustments: float
    warnings: list[str] = Field(default_factory=list)

File: crates/edi-core/src/schema.rs
//! Parsed EDI data structures. Field names are the contract the api-gateway
//! reads in `_clean_claim` / `_clean_denial`.

use serde::{Deserialize, Serialize};

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(default)]
pub struct ParsedPaymentInfo {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transaction_handling_code: Option<String>,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub payment_amount: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub credit_debit_flag: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payment_method: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payment_format: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payment_date: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub trace_number: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payer_identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub production_date: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payer_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payer_id_number: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payee_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payee_npi: Option<String>,
    /// PLB adjustments apply to the provider/payment, never a patient claim.
    pub provider_adjustments: Vec<ParsedProviderAdjustment>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(default)]
pub struct ParsedProviderAdjustment {
    pub provider_identifier: String,
    pub fiscal_period_date: Option<String>,
    pub adjustment_reason_code: String,
    pub reference_number: Option<String>,
    pub amount: f64,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(default)]
pub struct ParsedClaimAdjustment {
    #[serde(default = "default_level")]
    pub adjustment_group_code: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub reason_code: Option<String>,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub amount: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub quantity: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub remark_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub description: Option<String>,
    #[serde(skip_serializing_if = "is_default_level")]
    pub level: String,
}

fn default_level() -> String {
    "claim".to_string()
}

fn is_default_level(v: &str) -> bool {
    v == "claim"
}

impl Default for ParsedClaimAdjustment {
    fn default() -> Self {
        Self {
            adjustment_group_code: String::new(),
            reason_code: None,
            amount: 0.0,
            quantity: None,
            remark_code: None,
            description: None,
            level: "claim".to_string(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(default)]
pub struct ParsedServiceLine {
    pub line_number: i64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub procedure_qualifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cpt_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hcpcs_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub modifier_1: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub modifier_2: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub modifier_3: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub modifier_4: Option<String>,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub charge_amount: f64,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub paid_amount: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub revenue_code: Option<String>,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub units: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub allowed_amount: Option<f64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub service_date: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub control_number: Option<String>,
    pub remark_codes: Vec<String>,
}

impl Default for ParsedServiceLine {
    fn default() -> Self {
        Self {
            line_number: 0,
            procedure_qualifier: None,
            cpt_code: None,
            hcpcs_code: None,
            modifier_1: None,
            modifier_2: None,
            modifier_3: None,
            modifier_4: None,
            charge_amount: 0.0,
            paid_amount: 0.0,
            revenue_code: None,
            units: 1.0,
            allowed_amount: None,
            service_date: None,
            control_number: None,
            remark_codes: Vec::new(),
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(default)]
pub struct ParsedServiceLineDetail {
    pub service_line: ParsedServiceLine,
    pub adjustments: Vec<ParsedClaimAdjustment>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(default)]
pub struct ParsedClaim {
    pub claim_id: String,
    pub claim_status_code: String,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub total_charged: f64,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub total_paid: f64,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub patient_responsibility: f64,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub total_adjustment: f64,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub claim_filing_indicator: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payer_claim_control_number: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub facility_type_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub frequency_code: Option<String>,
    #[serde(default = "default_claim_type")]
    pub claim_type: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub patient_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub patient_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub date_of_birth: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider_npi: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub provider_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payer_name: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub payer_id_number: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub service_from: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub service_to: Option<String>,
    pub diagnosis_codes: Vec<String>,
    pub service_lines: Vec<ParsedServiceLineDetail>,
    pub claim_level_adjustments: Vec<ParsedClaimAdjustment>,
    pub remark_codes: Vec<String>,
    /// A payer that pays after this one: another subscriber on the 837 with a
    /// later responsibility sequence, or the 835's crossover carrier (NM1*TT).
    /// A patient-responsibility balance goes there before the patient.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_payer_name: Option<String>,
    /// `837_other_subscriber` or `835_crossover`.
    #[serde(skip_serializing_if = "Option::is_none")]
    pub next_payer_source: Option<String>,
}

fn default_claim_type() -> String {
    "professional".to_string()
}

impl Default for ParsedClaim {
    fn default() -> Self {
        Self {
            claim_id: String::new(),
            claim_status_code: String::new(),
            total_charged: 0.0,
            total_paid: 0.0,
            patient_responsibility: 0.0,
            total_adjustment: 0.0,
            claim_filing_indicator: None,
            payer_claim_control_number: None,
            facility_type_code: None,
            frequency_code: None,
            claim_type: "professional".to_string(),
            patient_id: None,
            patient_name: None,
            date_of_birth: None,
            provider_npi: None,
            provider_name: None,
            payer_name: None,
            payer_id_number: None,
            service_from: None,
            service_to: None,
            diagnosis_codes: Vec::new(),
            service_lines: Vec::new(),
            claim_level_adjustments: Vec::new(),
            remark_codes: Vec::new(),
            next_payer_name: None,
            next_payer_source: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize)]
#[serde(default)]
pub struct ParsedDenial {
    pub claim_id: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub service_line_number: Option<i64>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub cpt_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub hcpcs_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub modifier_1: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub modifier_2: Option<String>,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub charge_amount: f64,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub payment_amount: f64,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub adjustment_amount: f64,
    pub cagc: String,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub carc_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub rarc_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub denial_reason: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub denial_date: Option<String>,
}

impl Default for ParsedDenial {
    fn default() -> Self {
        Self {
            claim_id: String::new(),
            service_line_number: None,
            cpt_code: None,
            hcpcs_code: None,
            modifier_1: None,
            modifier_2: None,
            charge_amount: 0.0,
            payment_amount: 0.0,
            adjustment_amount: 0.0,
            cagc: String::new(),
            carc_code: None,
            rarc_code: None,
            denial_reason: None,
            denial_date: None,
        }
    }
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(default)]
pub struct ParsedTransactionMetadata {
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transaction_set_identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub transaction_set_control_number: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub implementation_convention_reference: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub interchange_sender_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub interchange_receiver_id: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub functional_group_code: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub version_identifier: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub element_separator: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub component_separator: Option<String>,
    #[serde(skip_serializing_if = "Option::is_none")]
    pub segment_terminator: Option<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize, Default)]
#[serde(default)]
pub struct Parsed835Response {
    pub metadata: ParsedTransactionMetadata,
    pub payment_info: ParsedPaymentInfo,
    pub claims: Vec<ParsedClaim>,
    pub denials: Vec<ParsedDenial>,
    pub total_claims: usize,
    pub total_denials: usize,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub total_charges: f64,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub total_payments: f64,
    #[serde(skip_serializing_if = "is_default_f64")]
    pub total_adjustments: f64,
    pub warnings: Vec<String>,
}

fn is_default_f64(v: &f64) -> bool {
    *v == 0.0
}
